Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+
+#### Detecting build from source
+
+The first parameter used by source-build will be a value indicating whether the build is being performed in "source only" mode. This
+means repos cannot use tools/packages/references that haven't already been built by the current build.
+
+`-p:DotNetBuildFromSource=true`
+
+Some repos build more projects than is necessary to build the .NET Core stack. For example, the `dotnet/roslyn` repo builds
+the C# compiler, which is needed, but it also builds Visual Studio extensions, which aren't needed. These VS extensions
+cannot be built without using an external dependency on tooling that isn't available when building from source.
+
+#### Available tools
+
+In order for repos to build, they need tools in order to do the build (ex. MSBuild). These tools are not generally available
+on all supported operating systems. As such, many repos today have the ability to pull these tools from the internet during the
+build.
+
+However, there are a couple problems with each repo downloading its own toolset.
+
+1. Different versions can cause inconsistencies
+1. It takes a non-trivial amount of time to download the toolset, if every repo does it during the complete .NET Core build,
+it will add a lot of overhead to the build.
+
+To solve this, source-build will have a set of common tools available that it will pass into each repos build.
+
+First, a [.NET Core SDK](https://www.microsoft.com/net/download/core#/sdk) will be available for repos to be able to invoke
+MSBuild, NuGet, etc. The path to the folder containing the `dotnet` executable will be available using the
+
+`-p:DotNetToolDir`
+
+property. This can be used to call `dotnet restore`, `dotnet build`, `dotnet publish`, `dotnet msbuild`, or any other CLI
+verb on the repo's projects.
+
+Second, a common set of tools used by the `dotnet` org repos is [BuildTools](https://github.com/dotnet/buildtools). This will
+be available using the
+
+`-p:DotNetBuildToolsDir`
+
+property. This can be used to invoke any current or future tool that is part of BuildTools. It is not required that each repo
+use BuildTools, but if the repo wants to, it can get the tools from this property.
+
+### Example
+
+Here is an example of the list of calls source-build will make on the `dotnet/corefx` repo:
+
+```
+/path/to/corefx/build.sh -t:Restore -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+/path/to/corefx/build.sh -t:Build -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+/path/to/corefx/build.sh -t:Publish -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+```
+
+> **Note:** Not all properties have been specified above. As each action gets designed in detail, more `DotNet` properties will be
+specified. For example, the `Publish` action above isn't getting told where to publish the assets to.
+
+### .rsp files
```

Comment:
I share Davis’ concern about response files and I think they would be great but hard to implement. I actually think it is a good idea as part of the spec to provide a set of scripts build.cmd/build.sh that will do the arg parsing for both and handle these properties and call msbuild on a build.proj in the root. That way these scripts can be exactly the same for all repo’s (once we figure out the native build parts) and share the same arg parsing.

Reply:
+1

Reply:
> I actually think it is a good idea as part of the spec to provide a set of scripts 

I agree it will be valuable to provide default/basic implementations for common things across repos.  However, I don't think it is valuable to write those implementations as part of the spec.  Let's get to a consensus on the basic contract before writing scripts.

Reply:
Sure didn't mean that they are part of the spec just that they are a default implementation of the spec once we finalize it.